platform/iphone/in_app_store.mm

```diff
@@ -116,167 +92,150 @@ - (void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProdu
+	}
+}
+
+#pragma mark - Helpers
+- (NSString *)localizedPrice:(SKProduct *)product {
+	NSNumberFormatter *numberFormatter = [self.class cachedPriceFormatter];
+	[numberFormatter setLocale:product.priceLocale];
+	return [numberFormatter stringFromNumber:product.price];
+}
+
++ (NSNumberFormatter *)cachedPriceFormatter {
+	static dispatch_once_t onceToken;
+	static NSNumberFormatter *gPriceFormatter = nil;
+	dispatch_once(&onceToken, ^{
+		gPriceFormatter = [[NSNumberFormatter alloc] init];
+		[gPriceFormatter setFormatterBehavior:NSNumberFormatterBehavior10_4];
+		[gPriceFormatter setNumberStyle:NSNumberFormatterCurrencyStyle];
+	});
+	return gPriceFormatter;
+}
 
 @end
 
+ProductsDelegate *gProductsDelegate = [[ProductsDelegate alloc] init];
+
+static inline Dictionary TransactionDictionary(SKPaymentTransaction *transaction) {
+	Dictionary ret;
+	ret["result"] = "ok";
+	ret["product_id"] = String::utf8([transaction.payment.productIdentifier UTF8String]);
+	if (transaction.transactionState == SKPaymentTransactionStatePurchased) {
+		ret["type"] = "purchase";
+		ret["transaction_id"] = String::utf8([transaction.transactionIdentifier UTF8String]);
+	} else if (transaction.transactionState == SKPaymentTransactionStateFailed) {
+		ret["type"] = "purchase";
+		ret["error"] = String::utf8([transaction.error.localizedDescription UTF8String]);
+	} else if (transaction.transactionState == SKPaymentTransactionStateRestored) {
+		ret["type"] = "restore";
+	}
+	return ret;
+}
+
+Error InAppStore::request_product_info(Variant p_params) {
+
+	Dictionary params = p_params;
+	ERR_FAIL_COND_V(!params.has("product_ids"), ERR_INVALID_PARAMETER);
+
+	PoolStringArray pids = params["product_ids"];
+	printf("************ request product info! %i\n", pids.size());
+
+	NSMutableSet *products = [NSMutableSet setWithCapacity:pids.size()];
+	for (int i = 0; i < pids.size(); i++) {
+		printf("******** adding %ls to product list\n", pids[i].c_str());
+		[products addObject:@(pids[i].utf8().get_data())];
+	}
+
+	// Create a products request from the identifiers for StoreKit
+	SKProductsRequest *request = [[SKProductsRequest alloc] initWithProductIdentifiers:[products copy]];
+	request.delegate = gProductsDelegate;
+	[request start];
+
+	return OK;
+}
+
+Error InAppStore::restore_purchases() {
+
+	printf("restoring purchases!\n");
+	[[SKPaymentQueue defaultQueue] restoreCompletedTransactions];
+
+	return OK;
+}
+
 Error InAppStore::purchase(Variant p_params) {
 
 	ERR_FAIL_COND_V(![SKPaymentQueue canMakePayments], ERR_UNAVAILABLE);
 	if (![SKPaymentQueue canMakePayments])
 		return ERR_UNAVAILABLE;
 
-	printf("purchasing!\n");
+	printf("attempting to purchase!\n");
+
 	Dictionary params = p_params;
 	ERR_FAIL_COND_V(!params.has("product_id"), ERR_INVALID_PARAMETER);
 
 	NSString *pid = [[[NSString alloc] initWithUTF8String:String(params["product_id"]).utf8().get_data()] autorelease];
 	SKPayment *payment = [SKPayment paymentWithProductIdentifier:pid];
```

Comment:
`paymentWithProductIdentifier:` is now deprecated and they recommend passing a pointer to the original `ProductIdentifier` that you received. We might be able to store them in an `NSDictionary` and key on the number passed back to GDScript.

Will require more thought.

Reply:
This also raises a warning on Travis CI currently:
```
platform/iphone/in_app_store.mm:276:34: warning: 'paymentWithProductIdentifier:' is deprecated: first deprecated in iOS 5.0 - Use +paymentWithProduct: after fetching the available products using SKProductsRequest [-Wdeprecated-declarations,29]
        SKPayment *payment = [SKPayment paymentWithProductIdentifier:pid];
                                         ^
/Applications/Xcode-9.4.1.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS11.4.sdk/System/Library/Frameworks/StoreKit.framework/Headers/SKPayment.h:22:1: note: 'paymentWithProductIdentifier:' has been explicitly marked deprecated here [2]
 + (id)paymentWithProductIdentifier:(NSString *)identifier NS_DEPRECATED_IOS(3_0, 5_0, "Use +paymentWithProduct: after fetching the available products using SKProductsRequest");
 ^
```

Reply:
> heavypoly sent an image
![image](https://matrix.org/_matrix/media/r0/thumbnail/matrix.org/kmqUTfroaLwIUjYyDwMNezsU?width=800&height=600)
Has anyone gotten this from Apple review? 
purchases seem to be working on a beta build...not sure whats going on

Deprecated in iOS 5 (released 8 years ago), it seems to be present in the 13.2 SDK (which is current version distributed with the latest stable Xcode), but might be finally removed in 13.3? (This is just a guess, I do not know what exactly is causing this, but I think this one is the primary candidate).